# Denovo_SNV_Analysis

This is a pipeline for the identification, characterization and prioritization of Denovo autosomal SNVs in whole genome sequencing (WGS) and RNA-sequencing data for patients with ID and / or congenital malformations. Raw WGS sequencing data needs to be processed with the [UMCU IAP pipeline](https://github.com/UMCUGenetics/IAP) before it can be used in this pipeline. The raw RNA-sequencing data needs to be processed by the [UMCU RNASeq pipeline](https://github.com/UMCUGenetics/RNASeq), before it can be used. For questions please feel free to contact me at: freek.manders@gmail.com. This pipeline was developed during my master internship in the [Cuppen group](https://www.umcutrecht.nl/en/Research/Research-centers/Center-for-Molecular-Medicine/Section-Genetics/Cuppen) at the Utrecht UMC.


## USAGE:
This pipeline works on the GRCh37 genome assembly.
To run the pipeline:
```bash
python Denovo_snv_analysis.py
```
Settings can be changed in the [Denovo_snv_analysis.ini](Denovo_snv_analysis.ini) file.
To run the pipeline with a different ini file:
```bash
python Denovo_snv_analysis.py --ini myini.ini
```
This pipeline is designed to run on a Sun Grid Engine system. To run the pipeline on a different system, you need to change the [Denovo_snv_analysis.py](Denovo_snv_analysis.py) script.


## Dependencies:
#### Core tools
- Sun Grid Engine
- Python >= 2.7.10
- R >= 3.4.1 (Bioconductor >= 3.6)

#### Bio tools loaded as modules in the Sun Grid Engine
- [bedtools >= 2.25.0](https://github.com/arq5x/bedtools2)
- [vcftools >= c7a7337](http://vcftools.sourceforge.net/)
- [tabix >= 0.2.6](http://www.htslib.org/doc/tabix.html)
- [bcftools >= 1.3](https://samtools.github.io/bcftools/)

#### Other Bio tools
- [GATK (genomeanalysis toolkit) >= 3.4-46](https://www.broadinstitute.org/gatk/)
- [SnpEff / SnpSift >= v4.1h](http://snpeff.sourceforge.net/)
- [denovogear >= 1.1.1](https://github.com/denovogear/denovogear)
- [position_effect](https://github.com/ibn-salem/position_effect)

#### Python packages
- argparse
- os
- sys
- re
- datetime
- timeit
- numpy
- pandas
- multiprocessing
- subprocess
- signal
- collections
- shutil

#### R packages
- biomaRt
- BSgenome.Hsapiens.UCSC.hg19
- cba
- circlize
- ComplexHeatmap
- dplyr
- gdata
- GenomicRanges
- ggplot2
- ggpubr
- ggrepel
- gridExtra
- Gviz
- karyoploteR
- MutationalPatterns
- optparse
- plyr
- RColorBrewer
- reshape2
- rtracklayer
- scales
- stringr
- VariantAnnotation
- VennDiagram

#### Datasets used when running the pipeline. Some of these may require some pre-processing
- [DANN scores](https://cbcl.ics.uci.edu/public_data/DANN/)
- [gnomAD](http://gnomad.broadinstitute.org/)
- [HMF-PON](https://www.hartwigmedicalfoundation.nl/) (Contact the HMF for acces to the data)
- [ROADMAP 15-state model](https://egg2.wustl.edu/roadmap/web_portal/chr_state_learning.html) (Download the coreMarks mnemonics.bed files for samples: E029, E070, E071, E081 and E082)
- [Ensembl Regulatory Builld](https://grch37.ensembl.org/index.html) (Download with biomart as: Ensembl Regulation: Human Regulatory Features. Tested on version 92. Make sure to download the version for GRCh37)
- [phastCons46way](http://hgdownload.cse.ucsc.edu/goldenPath/hg19/phastCons46way/vertebrate/) (Download the .wigFix.gz files)
- Age_parents (A tab delimited file with the format: sample_id, age_father, age_mother)
- Breakpoints (A file containing breakpoints of the SVs identified in this sample. For scripts to identify these SVs please contact me.)
- [Cancer signatures](http://cancer.sanger.ac.uk/cancergenome/assets/signatures_probabilities.txt)
- Output data from cartagenia. (This is only necessary when comparing the results of the pipeline to cartagenia
- A gene list generated by a script from the Cuppen group. Please contact me if you wish to use this script.
- An excell file containing the phenotypes of the patient. The first column should contain the patients ID, the third column should contain the phenotypes as HPO terms seperated by a ",".
- [Genes_to_phenotype](http://human-phenotype-ontology.github.io/) (tested on the all_sources_all_frequencies version from 2017-12-11.)
- [HPO OBO](http://human-phenotype-ontology.github.io/) (tested on the version from 2017-12-11.)
- [Entrez gene list](https://github.com/ibn-salem/position_effect)
- PCHiC data. Tested with the datasets from [Javierre BM et al.](https://www.sciencedirect.com/science/article/pii/S0092867416313228?via%3Dihub), [Paula Freire-Pritchett et al.](https://elifesciences.org/articles/21926), [Cairns J et al.](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0992-2) and [Adam J Rubin et al.](https://www.nature.com/articles/ng.3935)
